File: uni_modules/uts-any-rtc/utssdk/app-android/RtcEngineManager.kt
package uts.sdk.modules.utsAnyRtc

import android.content.Context
import android.view.TextureView
import com.ss.bytertc.engine.RTCEngine
import com.ss.bytertc.engine.VideoCanvas
import com.ss.bytertc.engine.VideoEncoderConfig
import com.ss.bytertc.engine.data.AudioRoute
import com.ss.bytertc.engine.data.CameraId
import com.ss.bytertc.engine.data.EngineConfig
import com.ss.bytertc.engine.data.RemoteStreamKey
import com.ss.bytertc.engine.data.StreamIndex

object RtcEngineManager {

    private var rtcEngine: RTCEngine? = null
    var isCapturingAudio: Boolean = false
        private set
    var isCapturingVideo: Boolean = false
        private set
    private var wasCapturingVideo: Boolean = false
    private var currentCameraId: CameraId = CameraId.CAMERA_ID_FRONT

    @Synchronized
    fun createEngine(context: Context, appId: String, handler: RtcEventProxy): RTCEngine? {
        if (rtcEngine != null) {
            return null
        }
        val config = EngineConfig()
        config.context = context.applicationContext
        config.appID = appId
        rtcEngine = RTCEngine.createRTCEngine(config, handler)
        return rtcEngine
    }

    @Synchronized
    fun destroyEngine() {
        stopAudioCapture()
        stopVideoCapture()
        RtcRoomManager.destroyAllRooms()
        RTCEngine.destroyRTCEngine()
        rtcEngine = null
        isCapturingAudio = false
        isCapturingVideo = false
        wasCapturingVideo = false
    }

    fun getEngine(): RTCEngine? {
        return rtcEngine
    }

    fun isInitialized(): Boolean {
        return rtcEngine != null
    }

    // Audio
    fun startAudioCapture() {
        rtcEngine?.startAudioCapture()
        isCapturingAudio = true
    }

    fun stopAudioCapture() {
        rtcEngine?.stopAudioCapture()
        isCapturingAudio = false
    }

    fun setDefaultAudioRoute(useSpeaker: Boolean) {
        rtcEngine?.setDefaultAudioRoute(
            if (useSpeaker) AudioRoute.AUDIO_ROUTE_SPEAKERPHONE
            else AudioRoute.AUDIO_ROUTE_EARPIECE
        )
    }

    // Video
    fun startVideoCapture() {
        rtcEngine?.startVideoCapture()
        isCapturingVideo = true
    }

    fun stopVideoCapture() {
        rtcEngine?.stopVideoCapture()
        isCapturingVideo = false
    }

    fun switchCamera(cameraId: CameraId) {
        currentCameraId = cameraId
        rtcEngine?.switchCamera(cameraId)
    }

    fun setLocalVideoCanvas(streamIndex: StreamIndex, canvas: VideoCanvas?) {
        rtcEngine?.setLocalVideoCanvas(streamIndex, canvas)
    }

    fun setRemoteVideoCanvas(roomId: String, uid: String, streamIndex: StreamIndex, canvas: VideoCanvas?) {
        val remoteStreamKey = RemoteStreamKey(roomId, uid, streamIndex)
        rtcEngine?.setRemoteVideoCanvas(remoteStreamKey, canvas)
    }

    fun setVideoEncoderConfig(config: VideoEncoderConfig) {
        rtcEngine?.setVideoEncoderConfig(config)
    }

    // Lifecycle
    fun onAppPause() {
        if (isCapturingVideo) {
            wasCapturingVideo = true
            rtcEngine?.stopVideoCapture()
            isCapturingVideo = false
        }
    }

    fun onAppResume() {
        if (wasCapturingVideo) {
            rtcEngine?.startVideoCapture()
            isCapturingVideo = true
            wasCapturingVideo = false
        }
    }

    // View creation helper
    fun createTextureView(context: Context): TextureView {
        return TextureView(context)
    }
}

File: uni_modules/uts-any-rtc/utssdk/app-ios/RtcEngineManager.swift
import Foundation
import UIKit
import AVFoundation
import BytePlusRTC

class RtcEngineManager {

    static let shared = RtcEngineManager()

    private var rtcVideo: ByteRTCEngine?
    var isCapturingAudio: Bool = false
    var isCapturingVideo: Bool = false
    private var wasCapturingVideo: Bool = false
    private var currentCameraId: ByteRTCCameraID = .front

    private init() {
        rtcVideo = nil
    }

    // MARK: - Engine Lifecycle

    func createEngine(appId: String, delegate: ByteRTCEngineDelegate?) -> ByteRTCEngine? {
        guard rtcVideo == nil else { return nil }
        let config = ByteRTCEngineConfig(); config.appID = appId; rtcVideo = ByteRTCEngine.createRTCEngine(config, delegate: delegate)
        return rtcVideo
    }

    func destroyEngine() {
        stopAudioCapture()
        stopVideoCapture()
        RtcRoomManager.shared.destroyAllRooms()
        ByteRTCEngine.destroyRTCEngine()
        rtcVideo = nil
        isCapturingAudio = false
        isCapturingVideo = false
        wasCapturingVideo = false
    }

    func getEngine() -> ByteRTCEngine? {
        return rtcVideo
    }

    func isInitialized() -> Bool {
        return rtcVideo != nil
    }

    // MARK: - Permission Check

    static func hasCameraPermission() -> Bool {
        let status = AVCaptureDevice.authorizationStatus(for: .video)
        if status == .notDetermined {
            // Trigger permission dialog, return false for now
            // Next call will return .authorized after user grants
            AVCaptureDevice.requestAccess(for: .video) { _ in }
            return false
        }
        return status == .authorized
    }

    static func hasMicrophonePermission() -> Bool {
        let status = AVCaptureDevice.authorizationStatus(for: .audio)
        if status == .notDetermined {
            AVCaptureDevice.requestAccess(for: .audio) { _ in }
            return false
        }
        return status == .authorized
    }

    // MARK: - Audio

    func startAudioCapture() {
        rtcVideo?.startAudioCapture()
        isCapturingAudio = true
    }

    func stopAudioCapture() {
        rtcVideo?.stopAudioCapture()
        isCapturingAudio = false
    }

    func setDefaultAudioRoute(useSpeaker: Bool) {
        rtcVideo?.setDefaultAudioRoute(useSpeaker ? .speakerphone : .earpiece)
    }

    // MARK: - Video

    func startVideoCapture() {
        rtcVideo?.startVideoCapture()
        isCapturingVideo = true
    }

    func stopVideoCapture() {
        rtcVideo?.stopVideoCapture()
        isCapturingVideo = false
    }

    func switchCamera(_ cameraId: ByteRTCCameraID) {
        currentCameraId = cameraId
        rtcVideo?.switchCamera(cameraId)
    }

    func setLocalVideoCanvas(streamIndex: ByteRTCStreamIndex, canvas: ByteRTCVideoCanvas?) {
        rtcVideo?.setLocalVideoCanvas(streamIndex, withCanvas: canvas)
    }

    func setRemoteVideoCanvas(roomId: String, uid: String, streamIndex: ByteRTCStreamIndex, canvas: ByteRTCVideoCanvas?) {
        let streamKey = ByteRTCRemoteStreamKey()
        streamKey.userId = uid
        streamKey.roomId = roomId
        streamKey.streamIndex = streamIndex
        rtcVideo?.setRemoteVideoCanvas(streamKey, withCanvas: canvas)
    }

    func setVideoEncoderConfig(_ config: ByteRTCVideoEncoderConfig) {
        rtcVideo?.setVideoEncoderConfig(config)
    }

    // MARK: - Lifecycle

    func onAppPause() {
        if isCapturingVideo {
            wasCapturingVideo = true
            rtcVideo?.stopVideoCapture()
            isCapturingVideo = false
        }
    }

    func onAppResume() {
        if wasCapturingVideo {
            rtcVideo?.startVideoCapture()
            isCapturingVideo = true
            wasCapturingVideo = false
        }
    }

    // MARK: - View Creation

    func createVideoView() -> UIView {
        return UIView()
    }
}
